Email verification

FabianRG1990 opened the pull request:
## ✅ Evidence of succesfull runs:

- `npx nx lint adrian-badilla` is executed without errors
- `npx nx test adrian-badilla` Pass all tests
- `npx nx build adrian-badilla` It compiles correctly
- `npx nx serve adrian-badilla` works as expected
- `Show case of feature implemented` works as expected
<img width="1524" height="932" alt="image" src="https://github.com/user-attachments/assets/3f595af0-5e95-4d9f-95b2-5678f280e44c" />
<img width="990" height="832" alt="image" src="https://github.com/user-attachments/assets/ba033cca-a84f-49e1-b5d8-48c038da3b0a" />

FabianRG1990 committed:
1: Add the component to my project and make it work by satisfying the dependencies

FabianRG1990 committed:
2: Make a route to see what's being rendered in the browser

FabianRG1990 committed:
feat: The error that SCSS flagged due to the weight was fixed; the test was also fixed.

FabianRG1990 committed:
bug: Integration of the login component comments

FabianRG1990 committed:
pasos 1,2,3 para EmailVerificationComponent

Plan de pasos para EmailVerificationComponent

1) Pasar el componente de la app vieja a la nueva

Comentar en el HTML las funciones que todavía no hemos creado y satisfacer las dependencias que hagan falta

2) Ver que el componente existe y se renderiza

Asegurarnos que la ruta y el componente sirven, sin lógica todavía.

Crear / revisar la ruta a la que Firebase va a mandar al usuario, algo tipo:

/auth/email-verification o algo similar según se ocupe.

Abrir esa ruta manualmente en el navegador (sin oobCode todavía).

Confirmar que se ven:

El título: "Verificación de cuenta".

El layout base (aunque los mensajes aún no estén conectados a nada).

Verificar que:

No hay errores de imports.

El componente carga sin problemas en la consola.

3) Crear/preparar la estructura del featureStore

Crear la base del feature e importarla al ts

Hacer un log para comprobar que se esta renderizando correctamente

Recordar importarlo al auth signal store

Llamarlo en el HTML para probar su funcionamiento

FabianRG1990 committed:
Merge branch 'main' into email-verification

FabianRG1990 committed:
Run: npx nx format

FabianRG1990 committed:
Merge branch 'email-verification' of https://github.com/cvandradg/adrian-badilla-ws into email-verification

FabianRG1990 committed:
paso 3.1

3.1 – Crear un rxMethod con un observable falso from('hola')

Ahora hacemos lo mismo, pero ya con rxMethod:

En el Feature de Email Verification (por ejemplo withEmailVerificationResources):

Crear un método con rxMethod<void> que:

Internamente haga from(['hola']).pipe(tap(x => console.log('probando log', x))).

Probar en el navegador:

Ver que aparece en consola: probando log hola.

Esto comprueba que:

El rxMethod se ejecuta correctamente.

La interconexión componente → store → rxMethod → consola está funcionando.

FabianRG1990 committed:
paso 3.2

3.2 – Reemplazar el observable falso por el real (solo logueando)

3.2.1- Entrar al feature withEmailVerificationResources.

3.2.2- Asegurarse de que el feature tenga:

withCustomCallState('emailVerification')

withProps

WithMethods

3.2.3- Dentro de withProps, inyectar el servicio:

firebaseAuthService: inject(FirebaseAuthService)

3.2.4- Dentro de withMethods, crear un método reactivo:

nombre sugerido: verifyEmailFromRouteTest

usando rxMethod<void>(...)

3.2.5- En el rxMethod, usar:

pipe(...)

dentro del pipe usar exhaustMap(...)

3.2.6- Dentro de exhaustMap:

crear un oobCode dummy → 'TEST_CODE'

llamar:

innerStore.firebaseAuthService.verifyEmail('TEST_CODE')

3.2.7- A ese observable aplicarle:

tapResponse({ … })

next: console.log('resultado verifyEmail (test):', res)

error: console.error('ERROR verifyEmail (test):', err)

3.2.8- No agregar todavía:

loading

success

error state
✔️ Solo logs.

3.2.9- En el componente:

seguir usando:

firebaseAuthStore = inject(firebaseAuthStore);

3.2.10- En el HTML:

llamar al método:

firebaseAuthStore.verifyEmailFromRouteTest()

3.2.11- Probar en navegador:

recargar

ejecutar el método

revisar la consola

3.2.12- Resultado esperado:

o resultado verifyEmail (test): ...

o ERROR verifyEmail (test): ...

FabianRG1990 committed:
bug: warning browser

cvandradg committed:
feat: duplication of keys removed

FabianRG1990 committed:
fix(auth): centralize dummy FirebaseAuthService and eliminate the warning for multiple injections

FabianRG1990 committed:
paso 3.2

3.2) Implementar y probar el envío del correo de verificación

Este paso es obligatorio antes de trabajar con oobCode

3.2.1) Confirmar que Firebase puede enviar correos de verificación.

3.2.2) Crear  el método en el servicio de Auth

3.2.3) Agregar un método en FirebaseAuthService que envíe el correo de verificación.

3.2.4) Agregar un método en FirebaseAuthService que permita obtener el usuario autenticado (sesión actual).

3.2.5) Crear el método sendEmailVerification en withEmailVerificationResources para enviar el correo.

Crear un método reactivo llamado sendEmailVerification.

El método debe tomar como fuente de verdad la sesión actual (usuario autenticado), no datos del componente.

El método debe usar esa sesión para intentar enviar el correo de verificación con Firebase.

El método debe imprimir logs claros en consola, en este orden:

“intentando enviar correo” (cuando se dispara el método)

“correo enviado” (si Firebase responde exitosamente)

“error enviando correo” (si Firebase responde con error)

Este paso se considera completado cuando al disparar el método:

se ve el log de “intentando enviar correo”

y luego se ve “correo enviado” o “error enviando correo” en consola.

3.2.6) Probar el envío en ambiente real

Hacer login con un usuario real.

Presionar el botón temporal para enviar el correo.

Validar el resultado técnico en consola

Confirmar que se imprimen los logs del método.

Confirmar que se ve éxito o error de Firebase (lo que sea, pero que responda).

Validar el resultado real en el correo

Revisar el inbox del usuario.

Confirmar que el correo llegó.

Confirmar que el correo trae un link de verificación.

Confirmar que el link lo lleve al la pagina asignada en firebase.

FabianRG1990 committed:
paso 5

5) Conectar estados del Store con la UI

5.1) Preparar el HTML para trabajar con señales del Store

Entrar al template del EmailVerificationComponent.

Identificar los 3 elementos visuales que vamos a controlar con estado:

Spinner / loading

Mensaje de éxito

Mensaje de error

Asegurar que el template pueda leer directamente del store:

emailVerificationLoading()

emailVerificationSuccess()

emailVerificationError()

5.2) Mostrar el spinner cuando emailVerificationLoading() sea true

Configurar el spinner para que solo sea visible cuando:

emailVerificationLoading() es true

Confirmar que cuando loading sea false, el spinner desaparece.

5.3) Mostrar el mensaje de éxito cuando el flujo termine correctamente

Configurar el mensaje de éxito para que solo se vea cuando:

emailVerificationSuccess() es true

y emailVerificationError() está vacío / null

Confirmar que el mensaje de éxito no aparece mientras loading está activo.

5.4) Mostrar el mensaje de error cuando exista un error

Configurar el mensaje de error para que solo se vea cuando:

emailVerificationError() tiene contenido (texto)

Confirmar que el mensaje de error no aparece si no hay error.

5.5) Validar que la UI refleja los estados en tiempo real (prueba real)

Entrar al componente usando un link real del correo (con oobCode).

Observar el comportamiento esperado:

Al iniciar → aparece el spinner (loading true)

Luego → desaparece el spinner y aparece:

éxito, si la verificación fue correcta

o error, si la verificación falló

5.7) Criterio de paso completado

Este paso se considera completado cuando:

El spinner aparece solo durante la llamada a Firebase.

El éxito se muestra únicamente cuando la verificación fue exitosa.

El error se muestra únicamente cuando Firebase devuelve un error.

La UI cambia correctamente en tiempo real en los 3 escenarios (válido, dañado, expirado).

FabianRG1990 committed:
6.1 pendiente

FabianRG1990 committed:
pasar el servicio a authStore

cvandradg committed:
feat: firebase service into store

FabianRG1990 committed:
test

FabianRG1990 committed:
faet: create lib dashboard

FabianRG1990 committed:
faet: login redirec to dashboar and run npx nx format

FabianRG1990 committed:
bug in routes the dashboard

FabianRG1990 committed:
paso 6

6) Navegación automática al dashboard después de verificación exitosa

6.1) Definir el objetivo del paso

Lograr que, después de aplicar el oobCode exitosamente, la app:

muestre loading mientras verifica

Si hay error, lo muestre en pantalla y NO navegue

Si es éxito, marque success y navegue automáticamente a /dashboard

6.2) Preparar el destino /dashboard (placeholder)

Crear un componente simple “Dashboard” (placeholder).

Registrar la ruta /dashboard en el router.

Probar manualmente en el navegador que:

/dashboard carga el componente sin errores

no hay errores de imports o rutas

Este placeholder sirve solo para validar la navegación y luego se reemplaza.

6.3) Conectar la navegación dentro del flujo real

Entrar al feature withEmailVerificationResources.

Identificar el método real que verifica el correo:

verifyEmailFromRoute

Confirmar que ese método ya hace:

emailVerificationSetLoading() antes de llamar Firebase

emailVerificationSetSuccess() si Firebase responde OK

emailVerificationSetError() si Firebase responde error

6.4) Navegar a /dashboard en el caso de éxito

Dentro de verifyEmailFromRoute, en tapResponse.next:

después de emailVerificationSetSuccess()

ejecutar navegación a /dashboard

Esto garantiza que:

solo si Firebase confirmó la verificación

la app continúa al dashboard

6.5) Probar el flujo completo con link real

Abrir un link real de verificación (con oobCode).

Confirmar el orden del flujo:

carga el componente

se activa loading (spinner)

Firebase responde:

Si éxito → success y navega a /dashboard

Si error → se muestra error en UI y NO navega

Confirmar que el placeholder de /dashboard se muestra correctamente.

6.6) Criterio de paso completado

Este paso se considera completado cuando:

/dashboard existe y carga sin errores

con link real válido:

se verifica el correo

se navega automáticamente a /dashboard

con link inválido/usado/expirado:

NO navega

se muestra el error en UI

FabianRG1990 committed:
paso 6

6.1 Contrato del flujo

Entrada: URL con oobCode.

Acción manual: botón “Verificar ahora”.

Estados UI:

loading: spinner + “Verificando…”

error: mensaje de error (Firebase)

success: mensaje “Cuenta verificada correctamente”

Navegación: manual con botón “Ir al dashboard” solo si success.

6.2 Feature: verifyEmailFromRoute (manual)

Al ejecutar:

setLoading

toma oobCode del signal

llama _verifyEmail(code)

setSuccess o setError

✅ Criterio de completado:

Con oobCode válido: success true, aparece botón dashboard.

Con oobCode inválido: error visible en UI.

6.3 Feature: goToDashboard (manual)

Navega a /dashboard.

✅ Criterio:

Solo se puede activar desde UI si success.

6.4 UI final EmailVerification

Mantener tu HTML tal cual (está correcto).

Confirmar que:

spinner se ve solo cuando loading

botón dashboard aparece solo con success

6.5 Pruebas del paso

Probar oobCode real → success + botón dashboard.

Probar oobCode falso → error visible.

FabianRG1990 committed:
paso 7

7) Flujo Register (crear cuenta + guiar al usuario)

7.0 Objetivo

Register crea cuenta.

Envío de verificación:

automático al crear cuenta (como ya lo tienes)

adicionalmente manual con botón “Reenviar verificación”

No navega a dashboard.

UI guía: “Revisa tu correo”.

7.1 Contrato del flujo

Register no verifica email.

Register no usa oobCode.

Register no hace login navigation.

EmailVerification es el que valida oobCode.

7.2 Feature: createAccount (punto de integración)

En tu código actual:

setLoading

_createAccount(creds)

_sendEmailVerification(cred.user)

setSuccess o setError

✅ Criterio:

Se crea usuario en Firebase.

Se intenta enviar correo.

registerSuccess() se pone true si todo ok.

7.3 UI Register: estados + guía + acciones

Tu HTML ya cumple:

loading / error / success

botón volver al login

botón reenviar

✅ Criterio:

Después de success: guía clara + botones visibles.

7.4 Reenviar verificación (manual)

Botón llama firebaseAuthStore.sendEmailVerification().

✅ Criterio:

Click → se intenta reenviar (y si sale too-many-requests, se muestra/queda log).

7.5 Pruebas del paso 8

Crear cuenta → success UI.

Reenviar → log “✅ correo enviado” o error de rate-limit.

FabianRG1990 committed:
fix: arreglo del bug que me permitia entrar al dashboard desde el login sin estar verificado

FabianRG1990 committed:
manejo de errores y limpieza del codigo

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/components/email-verification/email-verification.component.html:
```diff
@@ -0,0 +1,54 @@
+      "
+    />
+
+    @if ( firebaseAuthStore.emailVerificationSuccess() &&
```
Let's evaluate this, I don't think we need to have the 3 validations, how would it be success and have error at the same time?

FabianRG1990 replied:
<img width="671" height="182" alt="image" src="https://github.com/user-attachments/assets/99c3b42e-acd3-4b56-81c3-efaaf6b29f36" />

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/components/login/login.component.html:
```diff
@@ -16,52 +18,57 @@
+          [message]="firebaseAuthStore.loginError()!"
+          [display]="true"
+        />
+        } @if ( !firebaseAuthStore.loginLoading() &&
```
maybe we can evaluate a better way to avoid multiple ifs

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/components/register/register.component.scss:
```diff
@@ -56,7 +56,7 @@ button {
 .register-login-inputs {
   min-width: 22rem;
-
+  max-width: 22rem;
```
min with and max with are the same, maybe just use width?

FabianRG1990 replied:
<img width="332" height="56" alt="image" src="https://github.com/user-attachments/assets/df190699-5792-4443-a8b9-74791d193298" />

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/data-access/errors/auth-error.mapper.ts:
```diff
@@ -0,0 +1,63 @@
+  );
+}
+
+export function mapFirebaseAuthErrorToMessage(error: unknown): string {
```
why is this not a service or a store?

FabianRG1990 replied:
It’s intentionally not a service or store since it has no state or side effects. It’s just a pure mapper function. Using DI or state management here would add unnecessary complexity, so keeping it simple felt like the cleaner option.

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/data-access/stores/with-email-verification.feature.ts:
```diff
@@ -0,0 +1,92 @@
+          tap(() => innerStore.emailVerificationSetLoading()),
+          exhaustMap(() =>
+            store._getUserSession$().pipe(
+              filter((user): user is User => !!user),
```
why?

FabianRG1990 replied:
<img width="686" height="522" alt="image" src="https://github.com/user-attachments/assets/81b9c5e8-4685-49cb-89b1-d3b581a41478" />

Yep, you’re right — I checked it again and the issue was that filter(!!user) could leave the flow hanging when the session was null, which could cause the loading state to get stuck.

I switched it to take(1) + switchMap and now handle the null case explicitly, setting an error and completing the stream so the state always resolves properly.

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/data-access/stores/with-email-verification.feature.ts:
```diff
@@ -0,0 +1,92 @@
+            const code = innerStore._oobCode();
+            console.log('📩 oobCode leído desde la URL:', code);
+
+            return store._verifyEmail(code ?? '').pipe(
```
why do we need the ?? '' 
Signals must have a default value

FabianRG1990 replied:
<img width="430" height="167" alt="image" src="https://github.com/user-attachments/assets/4749febf-d70d-415b-9ce4-09d87b4deb10" />

<img width="693" height="493" alt="image" src="https://github.com/user-attachments/assets/ecd9656d-af8d-439d-8e2b-54c0e4e3bd32" />


Yes, I reviewed it more closely and saw the issue. Previously, we were using `code ?? ''` basically to force a string and avoid `null/undefined`, but that ultimately masked the real problem. If, for some reason, the `oobCode` wasn't available (due to the observable's timing or because the query parameter was missing), we ended up calling `_verifyEmail('')`, which isn't a valid value and could also leave the stream in a strange state (for example, the loading stage without a clear exit).

The solution was:

1. Now the `signal` has an explicit default value (`initialValue: null`), so it doesn't depend on `undefined` before the first `emit`.

2. Instead of using an empty string, we validated the case directly: if there was no `code`, we set a clear error and terminated the stream with `EMPTY`; if there was `code`, we called `_verifyEmail(code)` normally.

cvandradg commented on libs/adrian-badilla/frontend/auth/src/lib/data-access/stores/with-login.feature.ts:
```diff
@@ -1,49 +1,75 @@
-      firebaseAuthService: inject(FirebaseAuthService),
-    })),
-    withMethods((innerStore) => ({
+    withMethods((s) => ({
```
don't use single letters for parameters, use the whole word

FabianRG1990 replied:
<img width="658" height="465" alt="image" src="https://github.com/user-attachments/assets/76193a15-0838-432a-b210-001a347ebe7d" />

FabianRG1990 committed:
Update email-verification.component.html

FabianRG1990 committed:
Update register.component.scss

FabianRG1990 committed:
Fix for the sendEmailVerification method by comment on github

FabianRG1990 committed:
Fix for the verifyEmailFromRoute method by comment on githubmethod by comment on github

FabianRG1990 committed:
fix: Change the parameter of the login method from (s) to (innerStore)

FabianRG1990 committed:
bug: solucion de problema en los mensajes de error

FabianRG1990 committed:
fix: manejo de roran y redireccion de el link del correo

FabianRG1990 committed:
fix: solucion del guard dentro de la lib auth

FabianRG1990 committed:
bug: en el template despues de forzar un error el error se quedaba pegado y no podia hacer nada mas hasta que refrescara

FabianRG1990 committed:
feat: pruevas de cypress para login

FabianRG1990 committed:
feat: prueba de register con cypress

FabianRG1990 committed:
fix(workspace): resolve missing lib.routes references in auth and landing-page libs

FabianRG1990 committed:
feat: crear emulador de firebase para correr las pruebas sin datos reales

FabianRG1990 committed:
run: npx nx format

FabianRG1990 committed:
Merge branch 'email-verification' of https://github.com/cvandradg/adrian-badilla-ws into email-verification

FabianRG1990 committed:
fix: eliminar emulador de firebase

se determino que el emulador era la ruta incorrectar ya que si puede realiazar todas las pruevas pero todas se programan para que salgan como uno quiere y eso no esta ocurriendo como sucederia en un ambiente real donde si lo probamos con un usuario que no exista por ejemplo el error es jenuino no forzado

FabianRG1990 committed:
test(e2e): implement email verification flow with YOPmail

FabianRG1990 committed:
test(e2e): implement full auth flow with email verification and password reset via YOPmail

FabianRG1990 committed:
fix: Fix in the submit button, verification link, and redirection definition in the HTML

FabianRG1990 committed:
fix: test correction landing page

FabianRG1990 committed:
fix(layout): correct navbar responsive positioning and width behavior

FabianRG1990 committed:
Update auth.reset-password.cy.ts

FabianRG1990 committed:
feat(auth): add resend email verification flow with E2E coverage

FabianRG1990 committed:
Merge branch 'main' into email-verification
